core/station/impl/src/services/external_canister.rs

```diff
@@ -598,119 +608,133 @@ impl ExternalCanisterService {
+                        if !current_policies.change.contains(&policy_id) {
+                            print(format!(
+                                "Policy with id {} not found for external canister {}",
+                                Uuid::from_bytes(policy_id).hyphenated(),
+                                external_canister.canister_id.to_text()
+                            ));
+
+                            continue;
+                        }
 
-                        continue;
+                        self.request_policy_service.edit_request_policy(
+                            EditRequestPolicyOperationInput {
+                                policy_id,
+                                rule: Some(updated_change_policy.rule),
+                                specifier: None,
+                            },
+                        )?;
+                    }
+                    None => {
+                        self.request_policy_service.add_request_policy(
+                            AddRequestPolicyOperationInput {
+                                rule: updated_change_policy.rule,
+                                specifier: RequestSpecifier::ChangeExternalCanister(
+                                    ExternalCanisterId::Canister(external_canister.canister_id),
+                                ),
+                            },
+                        )?;
                     }
-
-                    self.request_policy_service.edit_request_policy(
-                        EditRequestPolicyOperationInput {
-                            policy_id,
-                            rule: Some(policy.rule),
-                            specifier: None,
-                        },
-                    )?;
-                }
-                None => {
-                    self.request_policy_service.add_request_policy(
-                        AddRequestPolicyOperationInput {
-                            rule: policy.rule,
-                            specifier: RequestSpecifier::ChangeExternalCanister(
-                                ExternalCanisterId::Canister(external_canister.canister_id),
-                            ),
-                        },
-                    )?;
                 }
             }
         }
 
-        // add or update the `Call` policies
-        for policy in input.calls {
-            match policy.policy_id {
-                Some(policy_id) => {
-                    if !current_policies.contains(&policy_id) {
-                        print(format!(
-                            "Policy with id {} not found for external canister {}",
-                            Uuid::from_bytes(policy_id).hyphenated(),
-                            external_canister.canister_id.to_text()
-                        ));
+        // manage the policies for the `Call` requests
+        if let Some(updated_call_policies) = input.calls {
+            let policies_to_remove = current_policies
```

Comment:
current_policies includes both change and calls so here we'd be removing all change policies (and vice-versa above), right?